core/src/test/scala/unit/kafka/server/ListGroupsRequestTest.scala

```diff
@@ -0,0 +1,236 @@
+    testListGroupsWithNewProtocol()
+  }
+
+  @ClusterTest(serverProperties = Array(
+    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "true"),
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "true"),
+    new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testListGroupsWithOldConsumerGroupProtocolAndNewGroupCoordinator(): Unit = {
+    testListGroupsWithOldProtocol()
+  }
+
+  @ClusterTest(clusterType = Type.ALL, serverProperties = Array(
+    new ClusterConfigProperty(key = "unstable.api.versions.enable", value = "false"),
+    new ClusterConfigProperty(key = "group.coordinator.new.enable", value = "false"),
+    new ClusterConfigProperty(key = "offsets.topic.num.partitions", value = "1"),
+    new ClusterConfigProperty(key = "offsets.topic.replication.factor", value = "1")
+  ))
+  def testListGroupsWithOldConsumerGroupProtocolAndOldGroupCoordinator(): Unit = {
+    testListGroupsWithOldProtocol()
+  }
+
+  private def testListGroupsWithNewProtocol(): Unit = {
+    if (!isNewGroupCoordinatorEnabled) {
+      fail("Cannot use the new protocol with the old group coordinator.")
+    }
+
+    // Creates the __consumer_offsets topics because it won't be created automatically
+    // in this test because it does not use FindCoordinator API.
+    createOffsetsTopic()
+
+    // Create the topic.
+    createTopic(
+      topic = "foo",
+      numPartitions = 3
+    )
+
+    for (version <- ApiKeys.LIST_GROUPS.oldestVersion() to ApiKeys.LIST_GROUPS.latestVersion(isUnstableApiEnabled)) {
+
+      // Join the consumer group. Note that we don't heartbeat here so we must use
+      // a session long enough for the duration of the test.
+      val (memberId1, _) = joinConsumerGroup("grp", true)
+
+      checkListedGroups(
+        groupId = "grp",
+        state = ConsumerGroupState.STABLE.toString,
+        statesFilterExpectingEmptyListedGroups = List(ConsumerGroupState.ASSIGNING.toString),
+        version = version
+      )
+
+      // Member 2 joins the group, triggering a rebalance.
+      val (memberId2, _) = joinConsumerGroup("grp", true)
+
+      checkListedGroups(
+        groupId = "grp",
+        state = ConsumerGroupState.RECONCILING.toString,
+        statesFilterExpectingEmptyListedGroups = List(ConsumerGroupState.STABLE.toString),
+        version = version
+      )
+
+      leaveGroup(
+        groupId = "grp",
+        memberId = memberId1,
+        useNewProtocol = true
+      )
+      leaveGroup(
+        groupId = "grp",
+        memberId = memberId2,
+        useNewProtocol = true
+      )
+
+      checkListedGroups(
+        groupId = "grp",
+        state = ConsumerGroupState.EMPTY.toString,
+        statesFilterExpectingEmptyListedGroups = List(ConsumerGroupState.RECONCILING.toString),
+        version = version
+      )
```

Comment:
I find this construct a bit weird because we mix two things into one. It may be better to separate. One way would be to list group and assert the returned list.